### PR TITLE
fix: Disallow type recursion during type declaration

### PR DESCRIPTION
**Pull request by MaxGraey**
- [x] I've read the contributing guidelines
- [x] I've added my name and email to the NOTICE file

### CODE REVIEW

File: src/parser.ts [hunk 1]
```diff
@@ -3499,6 +3521,13 @@ export class Parser extends DiagnosticEmitter {
         tn.skip(Token.BAR);
         let type = this.parseType(tn);
         if (!type) return null;
+        if (this.checkRecurseDefinitionForType(name.text, type)) {
+          this.error(
+            DiagnosticCode.Not_implemented_0,
+            tn.range(), "recursion in type aliases"
+          );
```
Comment (dcodeIO) [inline src/parser.ts:3528]: Would TS2456 fit here? `Type alias '{0}' circularly references itself.`

Comment (MaxGraey) [reply src/parser.ts:3528]: `TS2456` only appear for such case `type T = T | null` but success for `type T = () => T` for example.

Comment (dcodeIO) [reply src/parser.ts:3528]: There's also 2310: `"Type '{0}' recursively references itself as a base type."`. In general it seems that any of these two is preferable over a "not implemented", even if it doesn't match TS exactly.

**Review by dcodeIO — APPROVED** [review]

File: src/parser.ts (hunk 1 above)
Comment (dcodeIO) [reply src/parser.ts:3528]: Technically this makes sense, yeah, was mostly thinking about user expectation. Maybe you are right, leaving it up to you :)

File: src/parser.ts [hunk 2]
```diff
@@ -3499,6 +3521,13 @@ export class Parser extends DiagnosticEmitter {
         tn.skip(Token.BAR);
         let type = this.parseType(tn);
         if (!type) return null;
+        if (this.checkRecursiveTypeDeclaration(name.text, type)) {
+          this.error(
+            DiagnosticCode.Not_implemented_0,
+            tn.range(), "recursion in type aliases"
```
Comment (romdotdog) [inline src/parser.ts:3527]:
```suggestion
            tn.range(), "Recursion in type aliases"
```

File: src/parser.ts [hunk 3]
```diff
@@ -3478,6 +3478,33 @@ export class Parser extends DiagnosticEmitter {
...
+  private getRecursiveDepthForTypeDeclaration(
+    identifierName: string,
+    type: TypeNode,
+    depth: i32 = 0
+  ): i32 {
+    switch (type.kind) {
+      case NodeKind.NAMEDTYPE: {
+        if ((<NamedTypeNode>type).name.identifier.text == identifierName) {
+          return depth;
+        }
```
Comment (dcodeIO) [inline src/parser.ts:3490]: Should this perhaps also check type parameters if `type` is generic? For example
```ts
type A = B<A>;
```

Comment (MaxGraey) [reply src/parser.ts:3490]: DO you mean `type A = Array<A>`? It should be an `Not implementer` error while it support in TS

Comment (dcodeIO) [reply src/parser.ts:3490]: For example, yes. Seems this is not checked yet?

Comment (MaxGraey) [reply src/parser.ts:3500]: Yeah, This was not checked. Fix this scenario as well

**Pull request merged**